packages/commons/src/context/context.ts

```diff
@@ -0,0 +1,56 @@
+/* eslint-disable functional/immutable-data */
+import { AsyncLocalStorage } from "async_hooks";
+import { NextFunction, Request, Response } from "express";
+import { z } from "zod";
+import { AuthData } from "../auth/authData.js";
+import { readHeaders } from "../auth/headers.js";
+
+export const ctx = z.object({
+  authData: AuthData,
+  correlationId: z.string().uuid(),
+});
+
+export type AppContext = z.infer<typeof ctx>;
+
+const globalStore = new AsyncLocalStorage<AppContext>();
+const defaultAppContext = {
+  authData: {
+    userId: "",
+    organizationId: "",
+    userRoles: [],
+  },
+  correlationId: "",
+};
+
+export const getContext = (): AppContext => {
+  const context = globalStore.getStore();
+  return !context ? defaultAppContext : context;
+};
+
+export const globalContextMiddleware = (
+  _req: Request,
+  _res: Response,
+  next: NextFunction
+): void => {
+  globalStore.run(defaultAppContext, () => getContext());
```

Comment:
The access of local storage is done one time per request during logger-middleware execution, async local storage usages are managed between three middlewares: 

- globalContextMiddleware: init local storage with default data (in our case empty fields)  
- contextDataMiddleware: parse headers and read and writes metadata to local storage
- loggerMiddleware: create a logger instance and read metadata for custom format 

those middlewares are executed subsequently one time per HTTP request and the last of them reading metadata used  for a logger formatter. 
As mentioned in [Node documentation](https://nodejs.org/api/async_context.html#introduction)
> Async Local Storage propagate data throughout callbacks and promise chains, storing data throughout the lifetime of a web request or any other asynchronous duration, definitively it is similar to thread-local storage in other languages.

**In detail** 
Reading meta data was executed when logger format was created in [customFormat](https://github.com/pagopa/interop-be-monorepo/blob/d4a70084ea5454c78c4629fadecf97ac45f31e96/packages/commons/src/logging/loggerMiddleware.ts#L27) at logger initization.

Anyway, this line initialize local-storage, so could be more correct to return `defaultAppContext` also in callback instead using `getContext()` to read it.

Reply:
@galales I have investigate in more detail, and I found that callback provided to `winston.format.printf` was execute everytime when the Logger logging something, like as you have suggest in previous comment, anyway Async Local Storage should guarantee the storage data isolation for callback chains also in the async context, so during HTTP request's execution flow the init meta data remains the same until the last function was executed, also if performed in async way.  
From my perspective context-loss remains lower risk.

Reply:
An other possible solution create context object and provide it to ALL function wants logging, this approach require to rewriting signatures of all functions in chain, add necessary parameter in every logs execution (potentially this means that if we made some helper methods it needs to receives context data as parameter - otherwise correlation-id and metadata are lost during logging).   
The current solution solve the problem for correlation-id propagation between functions chain and avoid to use it as parameter every time in logging;

Reply:
This PR is crucial for others deployment, because implements some refactor in `commons` package that are will be required in all services, so I suggest to merge it to unlock other deployment base on it.
About possible performance issues I suggest to track it in jira as tech debt, then runs some stress tests to evaluate results.

Reply:
fine by me to approve this PR to avoid blockers, but yes I would please ask you to open the tech debt, because it's one of those things that may I have an important impact and that are very difficult to spot once the problem happens

Reply:
Sure 👍🏻  a performance analysis task created with [IMN-51](https://pagopa.atlassian.net/browse/IMN-51)